File: Tiles CLOUD/client-lib/tiles-client.js
'use strict';

var mqtt = require('mqtt');
var util = require('util');
var EventEmitter = require('events').EventEmitter;

var host = 'mqtt://test.mosquitto.org';
var tag = '[TILES Client]';

function TilesClient(username){
  this.mqttClient = null;
  this.isConnected = false;
  this.username = username;
}

TilesClient.prototype.__proto__ = EventEmitter.prototype;

TilesClient.prototype.setServerConnectionStatus = function(msg, isConnected){
  console.log(tag, msg);
  this.isConnected = isConnected;
}

TilesClient.prototype.connect = function(username) {
  this.mqttClient = mqtt.connect(host);
  this.setServerConnectionStatus('Connecting...', false);

  var that = this;

  this.mqttClient.on('connect', function(){
    that.setServerConnectionStatus('Successfully connected to server', true);
    that.mqttClient.subscribe('tiles/' + that.username + '/+');
    that.mqttClient.subscribe('tiles/' + that.username + '/+/active');
    that.emit('connect');
  });

  this.mqttClient.on('offline', function() {
    that.setServerConnectionStatus('Client gone offline', false);
  });

  this.mqttClient.on('close', function() {
    that.setServerConnectionStatus('Disconnected from server', false);
  });

  this.mqttClient.on('reconnect', function() {
    that.setServerConnectionStatus('A reconnect is started', false);
  });

  this.mqttClient.on('error', function(error) {
    console.log(tag, 'Error: ' + error);
  });

  this.mqttClient.on('message', function(topic, message) {
    var splitTopic = topic.split('/');
    var tileId = splitTopic[2];
    if (splitTopic[3] === 'active'){
      var tileChange = (message.toString() === 'true') ? 'tileRegistered' : 'tileUnregistered';
      that.emit(tileChange, tileId);
    } else {
      that.emit('receive', tileId, message);
    }
  });

  return this;
}

TilesClient.prototype.send = function(tileId, msg){
  if (this.isConnected){
    this.mqttClient.publish('tiles/' + this.username + '/' + tileId, msg);
  } else {
    console.log(tag, 'Client is not connected!');
  }
}

module.exports = TilesClient;

File: Tiles CLIENTS/java/src/main/java/no/ntnu/tiles/client/TilesClient.java
package no.ntnu.tiles.client;
import java.util.Random;

import org.eclipse.paho.client.mqttv3.IMqttActionListener;
import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
import org.eclipse.paho.client.mqttv3.IMqttToken;
import org.eclipse.paho.client.mqttv3.MqttAsyncClient;
import org.eclipse.paho.client.mqttv3.MqttCallback;
import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
import org.eclipse.paho.client.mqttv3.MqttException;
import org.eclipse.paho.client.mqttv3.MqttMessage;
import org.eclipse.paho.client.mqttv3.MqttPersistenceException;

/**
 * A client for event-driven (asynchronous) communication with the Tiles server
 */
public class TilesClient {

	private String host;

	private MqttAsyncClient mqttClient;
	private MqttCallbackHandler mqttCallbackHandler;

	private String username;
	private TilesCallback tilesCallback;

	/**
	 * @param tilesCallback A callback listener to use for asynchronous events.
	 * @param username Your username
	 * @param host Host to connect to. Can include port param by adding :<port> to string
	 * @see TilesCallback
	 */
	public TilesClient(TilesCallback tilesCallback, String username,String host){
		this.tilesCallback = tilesCallback;
		this.username = username;
		this.mqttCallbackHandler = new MqttCallbackHandler();
		this.host=host;
	}

	/**
	 * Connects to the server.
	 * {@link TilesCallback#connected()} is called when the client is successfully connected to the server.
	 */
	public void connect(){
		try {
			mqttClient = new MqttAsyncClient(host, "JavaClient_"+new Random().nextInt(99999));
	    	mqttClient.setCallback(mqttCallbackHandler);
	    	mqttClient.connect(null, mqttCallbackHandler);
		} catch (MqttException e) {
			e.printStackTrace();
		}
	}

	/**
	 * Sends a message to all clients listening for messages for the Tile with the provided ID
	 * @param tileId The ID of the Tile this message is intended for
	 * @param message A {@link String} representation of the message
	 */
	public void send(String tileId, String message){
		send(tileId, message.getBytes());
	}

	/**
	 * Sends a message to all clients listening for messages for the Tile with the provided ID
	 * @param tileId The ID of the Tile this message is intended for
	 * @param bytes The bytes to be sent
	 */
	public void send(String tileId, byte[] bytes){
		try {
			mqttClient.publish("tiles/cmd/"+this.username+"/"+tileId, new MqttMessage(bytes));
		} catch (MqttPersistenceException e) {
			e.printStackTrace();
		} catch (MqttException e) {
			e.printStackTrace();
		}
	}

	private class MqttCallbackHandler implements MqttCallback, IMqttActionListener {
		public void messageArrived(String topic, MqttMessage message) throws Exception {
			String[] splitTopic = topic.split("/");
			String deviceId = splitTopic[3];
			if (splitTopic.length > 4 && splitTopic[4].equals("active")){
				if (new String(message.getPayload()).equals("true")){
					tilesCallback.tileRegistered(deviceId);
				} else {
					tilesCallback.tileUnregistered(deviceId);
				}
			} else {
				tilesCallback.messageArrived(deviceId, message.getPayload());
			}
		}

		public void onSuccess(IMqttToken asyncActionToken) {
	    	try {
				mqttClient.subscribe(new String[]{"tiles/evt/" + username + "/+", "tiles/evt/" + username + "/+/active"}, new int[] {0, 0});
				tilesCallback.connected();
			} catch (MqttException e) {
				e.printStackTrace();
			}
		}

		public void connectionLost(Throwable cause) { }

		public void deliveryComplete(IMqttDeliveryToken token) { }

		public void onFailure(IMqttToken asyncActionToken, Throwable exception) { }
	}

}
